### ISSUE TITLE
[Refactor] Mock out BlockRepository to better unit test BlockStore logic.

### ISSUE THREAD

**Issue by fassadlr** [stratisproject/StratisBitcoinFullNode#296]
Find a solution to mock out the BlockRepository class so that we can use a fake/mocked out implementation during unit tests. 

There are a lot of unit tests which does not directly benefit from having a concrete BlockRepository instantiated and in those cases it would be more efficient/better to have it return mocked values.

Another solution is to create an "in memory" version of the BlockRepository so that we can still retain actual BlockRepository functionality but without writing to disk.

**Comment by fassadlr** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[BlockStore] Mock out BlockRepository to better unit test BlockStore logic.](https://trello.com/c/2Fl2hahw/160-blockstore-mock-out-blockrepository-to-better-unit-test-blockstore-logic)

**Comment by dangershony** [comment]
An in memory mock is an option also using Moq freamwork should work

**Comment by fassadlr** [comment]
@dangershony I see that BlockRepository implements IBlockRepository and that we do in fact Mock it out in certain unit tests already.

Just so that we are on the same page here, I would like to change the BlockStoreLoop to take in IBlockRepository so that we are in a position to inject an in memory repository. This repository will not write to disk. This will help us better test the Block Store Loop steps without depending on the actual repository.

I propose then that we move or create an integration test(s) to test the block store loop with the concrete implementation of BlockRepository.

Thoughts?

**Comment by fassadlr** [comment]
@dangershony Also, can I rename "BlockHash" in the BlockRepository to "Tip" or "LastStoredBlock"?

**Comment by dangershony** [comment]
Dont rename BlockHash its a property of the repo and no really a tip.

yeah sure on IBlockRepository  to the BlockStoreLoop definitely.
You dont necessarily need to create in an in memory implementation just use a mock, unless we have tests that mock a lot of blocks.

**Comment by fassadlr** [comment]
> Dont rename BlockHash its a property of the repo and no really a tip.

Cool!

> yeah sure on IBlockRepository to the BlockStoreLoop definitely.
I've changed BlockStoreLoop  to take in IBlockRepository now but neede to change BlockStoreStats as well to take in BlockStoreLoop.

> You dont necessarily need to create in an in memory implementation just use a mock, unless we have tests that mock a lot of blocks.

I've implemented an InMemory version that only deals with blocks at the moment. I think it works really well and its a great plus not having to write to disk all the time for unit tests. I've added an integration tests which tests the concrete implementation of IBlockRepository with the BlockStoreLoop. My issue with Mocking everything is that we'll need to set up a mock of IBlockRepository with subsequent returns for adding and deleting blocks, every time we set up a unit test. With an in memory version, you just "new" an in memory version and start working with it out the box.

How about I do a PR (or you pull my branch), check the code and see how you feel about it? I like it ;)
